aerleon/api.py

```diff
@@ -0,0 +1,352 @@
+! $Revision:$
+no ip access-list extended test-filter
+ip access-list extended test-filter
+ remark $Id:$
+
+
+ remark deny-to-reserved
+ deny ip any 0.0.0.0 0.255.255.255
+ deny ip any 10.0.0.0 0.255.255.255
+
+
+ remark deny-to-bogons
+ deny ip any 192.0.0.0 0.0.0.255
+ deny ip any 192.0.2.0 0.0.0.255
+
+
+ remark allow-web-to-mail
+ permit ip any host 200.1.2.4
+ permit ip any host 200.1.2.5
+
+exit
+```
+"""
+import copy
+import multiprocessing
+import pathlib
+import sys
+import typing
+
+from absl import logging
+
+from aerleon.aclgen import ACLParserError, ACLGeneratorError, RenderACL, WriteFiles, WriteList
+
+from aerleon.lib import aclgenerator
+from aerleon.lib import naming
+from aerleon.lib import plugin_supervisor
+from aerleon.lib import policy
+from aerleon.lib import policy_builder
+
+
+def Generate(
+    policies,
+    definitions,
+    output_directory: pathlib.Path = None,
+    optimize=False,
+    shade_check=False,
+    exp_info=2,
+) -> "dict[str, str]":
+
+    context = multiprocessing.get_context()
+    return _Generate(
+        policies,
+        definitions,
+        context,
+        output_directory,
+        optimize,
+        shade_check,
+        exp_info,
+    )
+
+
+def _Generate(
+    policies,
+    definitions: naming.Naming,
+    context: multiprocessing.context.BaseContext,
+    output_directory=None,
+    optimize=False,
+    shade_check=False,
+    exp_info=2,
+    max_renderers=1,
+) -> "dict[str, str]":
+
+    # thead-safe list for storing files to write
+    manager: multiprocessing.managers.SyncManager = context.Manager()
+    write_files: WriteList = manager.list()
+    errors: list = manager.list()
+    generated_configs: dict = manager.dict()
+
+    if max_renderers == 1:
+        for input_policy in policies:
+            _GenerateACL(
+                input_policy,
+                definitions,
+                write_files,
+                generated_configs,
+                output_directory,
+                optimize,
+                shade_check,
+                exp_info,
+            )
+    else:
+        pool = context.Pool(processes=max_renderers)
+        async_results: list[multiprocessing.pool.AsyncResult] = []
+        for input_policy in policies:
+            async_result = pool.apply_async(
+                _GenerateACL,
+                args=(
+                    input_policy,
+                    definitions,
+                    write_files,
+                    generated_configs,
+                    output_directory,
+                    optimize,
+                    shade_check,
+                    exp_info,
+                ),
+            )
+            async_results.append(async_result)
+        pool.close()
+        pool.join()
+        for result in async_results:
+            try:
+                result.get()
+            except (ACLParserError, ACLGeneratorError) as e:
+                logging.warning('\n\nerror encountered in rendering process:\n%s\n\n', e)
```

Comment:
Would f strings be better here?

Reply:
Yes, keeping this code identical to aclgen.py for the moment until we can consolidate the error handling for both cases, see #150 .